File: frontend/src/rpc/client/data/perusahaan.js
import TDataPerusahaanService from '@/rpc/gen/TDataPerusahaanService';
import { TUserRole } from '@/rpc/gen/user.user.types_types';
import { TCrudClient } from '@/rpc/client/crud';


class TDataPerusahaanClient extends TCrudClient{
	constructor(stores=null){
		super(
			stores, 
			TDataPerusahaanService, 
			'/api/data/perusahaan',
			{
				fetch: TUserRole.ADMIN_UTAMA,
				create: TUserRole.ADMIN_UTAMA,
				delete: TUserRole.SUPER_ADMIN,
			},
			{
				name: TUserRole.SUPER_ADMIN,
				enabled: TUserRole.ADMIN_UTAMA,
			}
		);
	}
}

const dataPerusahaanClient = new TDataPerusahaanClient();

export { TDataPerusahaanClient, dataPerusahaanClient }
export default dataPerusahaanClient

File: backend/rpc/handler/data/perusahaan.py
from rpc.gen.data.perusahaan.services import TDataPerusahaanService
from rpc.gen.data.perusahaan.structs.ttypes import TPerusahaan

from rpc.gen.user.user.types.ttypes import TUserRole
from ..crud import TCrudServiceHandler

from models import models

class TDataPerusahaanServiceHandler(TDataPerusahaanService.Iface, TCrudServiceHandler):
    def __init__(self):
        TDataPerusahaanService.Iface.__init__(self)
        TCrudServiceHandler.__init__(
            self,
            "perusahaan", 
            TPerusahaan,
            actions={
                "get": TUserRole.ADMIN_BIASA,
                "create": TUserRole.ADMIN_UTAMA,
                "fetch": TUserRole.ADMIN_BIASA,
                "delete": TUserRole.SUPER_ADMIN,
            },
            setters={
                "name": TUserRole.ADMIN_UTAMA,
                "enabled": TUserRole.ADMIN_UTAMA,
            },
            require_role=False
        )

